### PR TITLE
Allow delete multiple resources with the same name

### LINKED ISSUE #4475: Kubectl should allow `kubectl delete service,rc <name>`
For when you have service / rc / other things with the same name.

### PR DESCRIPTION
**Pull request by feihujiang**
Fixes #4475

### CODE REVIEW

**Comment by feihujiang** [comment]
@smarterclayton @bgrant0607 @JanetKuo Could you please review this PR?

**Comment by piosz** [comment]
@k8s-bot ok to test

File: pkg/kubectl/resource/builder.go [hunk 1]
```diff
@@ -532,34 +532,30 @@ func (b *Builder) visitorResult() *Result {
 		if len(b.resources) == 0 {
 			return &Result{singular: isSingular, err: fmt.Errorf("you must provide a resource and a resource name together")}
 		}
-		if len(b.resources) > 1 {
-			return &Result{singular: isSingular, err: fmt.Errorf("you must specify only one resource")}
-		}
 
+		visitors := []Visitor{}
```
Comment (smarterclayton) [inline pkg/kubectl/resource/builder.go]: I would recommend instead that when you have `a,b,c d` in ResourceTypeOrName that you convert them to resource tuples (which already handle this case).  Then you can remove this entire block (I think tuples is a strict superset of names at this point).

Comment (smarterclayton) [inline pkg/kubectl/resource/builder.go]: By tuples I mean the `a/d` form - so treat `a,b,c d` as a transform to `a/d b/d c/d`

**Comment by feihujiang** [comment]
@smarterclayton I changed this according to your suggestion, but I think we should convert them to resource tuples in RunDelete function, not  in ResourceTypeOrName function. Because  ResourceTypeOrName is invoked in many place, not only invoked by delete. What's your point?

**Comment by smarterclayton** [comment]
We shouldn't support a particular resource builder syntax in just one command (only in RunDelete).  If we want the ability to `kubectl delete foo,bar,baz a`, we also want `kubectl get foo,bar,baz a` to work.  So ResourceTypeOrName is the right place to do that.

**Comment by feihujiang** [comment]
@smarterclayton  OK, I see. And another question, should we support  "kubectl delete foo,bar  a  b", as a transform to foo/a foo/b bar/a bar/b?

**Comment by smarterclayton** [comment]
I don't see why we couldn't.  When you use ResourceTypeNameOrArgs you're
implicitly giving over every argument to the call, so being consistent with
user expectations is fine.

On Thu, Aug 20, 2015 at 10:55 AM, feihujiang <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> OK, I see. And
> another question, should we support "kubectl delete foo,bar a b", as a
> transform to foo/a foo/b bar/a bar/b?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12900#issuecomment-133039944>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by feihujiang** [comment]
@smarterclayton I changed, please review.

File: pkg/kubectl/resource/builder.go [hunk 2]
```diff
@@ -256,6 +256,23 @@ func (b *Builder) SelectAllParam(selectAll bool) *Builder {
 // When two or more arguments are received, they must be a single type and resource name(s).
 // The allowEmptySelector permits to select all the resources (via Everything func).
 func (b *Builder) ResourceTypeOrNameArgs(allowEmptySelector bool, args ...string) *Builder {
```
Comment (smarterclayton) [inline pkg/kubectl/resource/builder.go]: Can you add a test case in resource/builder_test.go to verify the new cases we've added (`1,2,3 a b` and `1,2 a`)?

**Comment by feihujiang** [comment]
@smarterclayton I added a test case in resource/builder_test.go, please review.

**Comment by smarterclayton** [comment]
LGTM

**Pull request merged**